File: solutions/exercise-6/src/main/java/io/miragon/training/application/service/RegisterMembershipService.java
package io.miragon.training.application.service;

import io.miragon.training.application.port.inbound.RegisterMembershipUseCase;
import io.miragon.training.application.port.outbound.MembershipProcess;
import io.miragon.training.application.port.outbound.MembershipRepository;
import io.miragon.training.domain.Age;
import io.miragon.training.domain.Email;
import io.miragon.training.domain.Membership;
import io.miragon.training.domain.MembershipId;
import io.miragon.training.domain.Name;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Transactional
public class RegisterMembershipService implements RegisterMembershipUseCase {

    private static final Logger log = LoggerFactory.getLogger(RegisterMembershipService.class);

    private final MembershipRepository repository;
    private final MembershipProcess process;

    public RegisterMembershipService(MembershipRepository repository, MembershipProcess process) {
        this.repository = repository;
        this.process = process;
    }

    @Override
    public MembershipId register(Command command) {
        log.info("Registering membership for {}", command.email());
        var membership = new Membership(new Email(command.email()), new Name(command.name()), new Age(command.age()));
        repository.save(membership);
        process.startProcess(membership);
        return membership.id();
    }
}

File: exercises/src/main/kotlin/io/miragon/training/application/service/RegisterMembershipService.kt
package io.miragon.training.application.service

import io.github.oshai.kotlinlogging.KotlinLogging
import io.miragon.training.application.port.inbound.RegisterMembershipUseCase
import io.miragon.training.application.port.outbound.MembershipProcess
import io.miragon.training.application.port.outbound.MembershipRepository
import io.miragon.training.domain.Age
import io.miragon.training.domain.Email
import io.miragon.training.domain.Membership
import io.miragon.training.domain.MembershipId
import io.miragon.training.domain.Name
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
class RegisterMembershipService(
    private val repository: MembershipRepository,
    private val process: MembershipProcess,
) : RegisterMembershipUseCase {

    private val log = KotlinLogging.logger {}

    override fun register(command: RegisterMembershipUseCase.Command): MembershipId {
        log.info { "Registering membership for ${command.email}" }

        // TODO Aufgabe 1:
        //  1. Erstelle ein Membership-Objekt aus dem Command (Email, Name, Age)
        //  2. Speichere es mit repository.save(...)
        //  3. Starte den Prozess mit process.startProcess(...)
        //  4. Gib die MembershipId zurück
        TODO("Aufgabe 1: Implementiere die Membership-Registrierung")
    }
}
